webui/react/src/stores/projects.tsx

```diff
@@ -49,6 +53,96 @@ class ProjectStore {
       return projects ? Loaded(projects) : NotLoaded;
     });
   }
+
+  public deleteProject(id: number) {
+    let deleted: Project | undefined;
+    this.#projects.update((prev) =>
+      prev.withMutations((map) => {
+        deleted = map.get(id);
+        return map.delete(id);
+      }),
+    );
+    if (deleted) {
+      this.#projectsByWorkspace.update((prev) =>
+        prev.withMutations((map) => {
+          if (deleted) {
+            const ws = map.get(deleted.workspaceId.toString());
+            if (ws) {
+              remove(ws, (p) => p.id === id);
+              map.set(deleted.workspaceId.toString(), [...ws]);
+            }
+          }
+          return map;
+        }),
+      );
+    }
+  }
+
+  #upsert(op: Project, np: Project) {
+    op.name = np.name;
+    op.description = np.description;
+    op.archived = np.archived;
+  }
+
+  public upsertProject(p: Project) {
+    let project: Project | undefined;
+    let projectInWs: Project | undefined;
+
+    this.#projects.update((prev) =>
+      prev.withMutations((map) => {
+        project = map.get(p.id);
+        if (project) {
+          this.#upsert(project, p);
+        } else {
+          map.set(p.id, { ...p });
+        }
+        return map;
+      }),
+    );
+    this.#projectsByWorkspace.update((prev) =>
+      prev.withMutations((map) => {
+        projectInWs = find(map.get(p.workspaceId.toString()), (tp) => tp.id === p.id);
+        if (projectInWs) {
+          // The workspaceId has not changed, just update
+          this.#upsert(projectInWs, p);
+        } else {
+          // The workspaceId has changed, add to the new workspace and remove from the old workspace
+          const ws = map.get(p.workspaceId.toString());
+          if (ws) {
+            ws.push(p);
+            map.set(p.workspaceId.toString(), [...ws]);
+          }
+          if (project) {
+            const ows = map.get(project.workspaceId.toString());
+            if (ows) {
+              remove(ows, (op) => op.id === p.id);
+              map.set(project.workspaceId.toString(), [...ows]);
+            }
+          }
+        }
+        return map;
+      }),
+    );
+  }
+
+  public subscribe(stream: Stream, spec: ProjectSpec) {
+    stream.subscribe(spec);
+  }
 }
 
 export default new ProjectStore();
+
+export const mapStreamProject = (p: StreamContent): Project => ({
```

Comment:
question: what is the long term plan for the streaming updates migration?

I'm concerned that if we try to map streaming types to the old proto types, we're gonna run into serious problems where the proto types were poorly defined in the first place and the mapping is really difficult.  Like with Checkpoints, where the data for a Checkpoint is derived from like 7 different tables (7 different streamable entities).

How should we resolve that moving forward?